mdp/nodes/openai_gym_nodes.py

```diff
@@ -0,0 +1,293 @@
+    For more information about the Gym library refer to the
+    documentation provided at https://gym.openai.com/
+
+    This node is a non-trainable node. The node's execute call takes
+    an action array as an input and returns a flattened array of
+    (current observation (observation), future observation, action, reward, done flag).
+
+    The node's metaclass is selected to be an OnlineNode instead of a Node
+    in order to enforce a shared numx_rng between the node and the gym's
+    environment.
+
+    The node also provides additional utility methods:
+    'stop_rendering' - closes gym's rendering window if initialized.
+
+    'get_environment_samples' - generates a required number of output
+    samples for randomly selected actions, starting from the current env state.
+    Note that the env state changes after the method is called.
+    This is equivalent to an execute call without any input.
+
+    'get_random_actions' - returns a random set of valid actions
+    within the environment.
+
+    'reset' - resets the environment.
+
+
+    **Instance variables of interest**
+
+      ``self.observation_dim / self.action_dim``
+         Flattened observation  / action dimension
+
+      ``self.observation_shape / self.action_shape``
+         The original observation shape / action shape. Eg. (100,80,3) for an RGB image.
+
+      ``self.observation_type / self.action_type``
+         Discrete or continuous observation / action space.
+
+      ``self.observation_lims / self.action_lims``
+         Upper and lower bounds of observation / action space.
+
+      ``self.n_observations / self.n_actions``
+         Number of observations or actions for discrete types.
+
+    """
+
+    def __init__(self, env_name, render=False, render_interval=1, auto_reset=True, dtype=None, numx_rng=None):
+        """
+        env_name - Registered gym environment name. Eg. "MountainCar-v0"
+        render - Enable or disable rendering. Disabled by default.
+        render_interval - Number of execution steps delay between rendering
+        auto_reset - Automatically resets the environment if gym env's done is True.
+        """
+        super(GymNode, self).__init__(input_dim=None, output_dim=None, dtype=dtype, numx_rng=None)
+
+        self._env_registry = gym.envs.registry.env_specs
+        if env_name in self._env_registry.keys():
+            self._env_name = env_name
+        else:
+            similar_envs_str = '(' + ', '.join(self._get_similar_env_names(env_name)) + ')'
+            raise mdp.NodeException(
+                "Unregistered environment name. Are you looking for any of these?: \n%s" % similar_envs_str)
+        self.env = gym.make(self.env_name)
+
+        self.render = render
+        self.auto_reset = auto_reset
+
+        # set a shared numx_rng
+        self.numx_rng = numx_rng
+
+        # get observation dims and shape
+        if isinstance(self.env.observation_space, spaces.discrete.Discrete):
```

Comment:
It would be nice to use polymorphism instead of explicit type checks. But I don't yet understand the code deep enough to tell if this would be feasible.